src/Microsoft.Health.Fhir.Core/Features/Operations/ConvertData/ContainerRegistryTemplateProvider.cs

```diff
@@ -73,7 +79,31 @@ async Task<string> TokenEntryFactory(ICacheEntry entry)
 
             try
             {
-                var provider = _templateCollectionProviderFactory.CreateTemplateCollectionProvider(request.TemplateCollectionReference, accessToken);
+                /*
+                   First try to get the template provider from the cache.
+                   If template provider is not in cache, then limit the number of threads that can create new providers to 1, and then create a new provider.
+                   This is to limit the amount of providers created in a multi-threaded scenario, so that we do not have multiple providers pulling the same image from the ACR.
+                */
+                ITemplateCollectionProvider provider = _templateProviderCache.Get(request.TemplateCollectionReference) as TemplateCollectionProvider;
```

Comment:
Thanks for the review. Yeah it would be better to cast to the interface `ITemplateCollectionProvider` rather than over casting to `TemplateCollectionProvider`. Updated the code to cast to `ITemplateCollectionProvider`